platform/darwin/src/NSExpression+MGLAdditions.mm

```diff
@@ -27,12 +29,39 @@ @implementation NSExpression (MGLAdditions)
         return { std::string([(NSString *)value UTF8String]) };
     } else if ([value isKindOfClass:NSNumber.class]) {
         NSNumber *number = (NSNumber *)value;
-        if((strcmp([number objCType], @encode(int))) == 0) {
-            return { number.intValue };
-        } else if ((strcmp([number objCType], @encode(double))) == 0) {
-            return { number.doubleValue };
-        } else if ((strcmp([number objCType], @encode(char))) == 0) {
-            return { number.boolValue };
+        if ((strcmp([number objCType], @encode(char)) == 0) ||
+            (strcmp([number objCType], @encode(BOOL)) == 0)) {
+            // char: 32-bit boolean
+            // BOOL: 64-bit boolean
+            return { (bool)number.boolValue };
+        } else if (strcmp([number objCType], @encode(short))     == 0 ||
+                   strcmp([number objCType], @encode(int))       == 0 ||
+                   strcmp([number objCType], @encode(long))      == 0 ||
+                   strcmp([number objCType], @encode(long long)) == 0 ||
+                   strcmp([number objCType], @encode(NSInteger)) == 0) {
+            // Triggered for most non-boolean whole numbers as converted
+            // by NSExpression regardless of signedness. NSNumber seems
+            // to try to save space where possible.
+            return { (int64_t)number.integerValue };
+        } else if (strcmp([number objCType], @encode(unsigned char))      == 0 ||
+                   strcmp([number objCType], @encode(unsigned short))     == 0 ||
+                   strcmp([number objCType], @encode(unsigned int))       == 0 ||
+                   strcmp([number objCType], @encode(unsigned long))      == 0 ||
+                   strcmp([number objCType], @encode(unsigned long long)) == 0 ||
+                   strcmp([number objCType], @encode(NSUInteger))         == 0) {
+            // Triggered for very large numbers, e.g. NSUIntegerMax. Handling
+            // all unsigned types here anyway for consistency.
+            return { (uint64_t)number.unsignedIntegerValue };
+        } else if (strcmp([number objCType], @encode(double)) == 0) {
+            // Double values on all platforms are interpreted precisely.
+            return { (double)number.doubleValue };
+        } else if (strcmp([number objCType], @encode(float)) == 0) {
+            // Float values when taken as double introduce precision problems,
+            // so warn the user to avoid them. This would require them to
+            // explicitly use -[NSNumber numberWithFloat:] arguments anyway.
+            // We still do this conversion in order to provide a valid value.
+            mbgl::Log::Warning(mbgl::Event::Style, "float values are converted to double and can introduce imprecision; please use double values explicitly in predicate arguments");
```

Comment:
Why `mbgl::Log`? Since this is in platform-specific code, we can issue an `NSLog()` directly. We should also stick it in a `dispatch_once()`, like we did in #6163, to avoid potentially spamming the console.